Prevent consecutive ::cs_main locks

hebasto opened the pull request:
This PR prevents consecutive `::cs_main` locks in the following functions:
- `FlushStateToDisk()` that could happen when it is called from:
  - `AcceptToMemoryPoolWithTime()`
  - `CChainState::DisconnectTip()`
  - `CChainState::ConnectTip()`
  - `CChainState::AcceptBlock()`
  - `CChainState::RewindBlockIndex()`
- UnloadBlockIndex()
- CChainState::ReplayBlocks()

hebasto committed:
Prevent consecutive cs_main locks in FlushStateToDisk

MarcoFalke commented on src/validation.cpp:
```diff
@@ -2234,7 +2234,6 @@ bool CChainState::FlushStateToDisk(
     FlushStateMode mode,
     int nManualPruneHeight)
 {
-    LOCK(cs_main);
```
Removed locks are generally replaced with `AssertLockHeld(cs_main);` and the corresponding compiler annotations

hebasto replied:
[Updated](https://github.com/bitcoin/bitcoin/pull/18639#issuecomment-613708604).

MarcoFalke replied:
Please also add the annotation in the header

hebasto replied:
> Please also add the annotation in the header
> 
> EXCLUSIVE_LOCKS_REQUIRED(cs_main)

It is already done: https://github.com/bitcoin/bitcoin/blob/fe187562ef9765942d7c7d801e04d70d294d5aef/src/validation.h#L671-L675

hebasto commented:
Updated 9b4b3c9442f5b2ccdce8d70161a355aec5402cef  -> fe187562ef9765942d7c7d801e04d70d294d5aef ([pr18639.01](https://github.com/hebasto/bitcoin/commits/pr18639.01) -> [pr18639.02](https://github.com/hebasto/bitcoin/commits/pr18639.02), [diff](https://github.com/hebasto/bitcoin/compare/pr18639.01..pr18639.02)):

- addressed @MarcoFalke's [comment](https://github.com/bitcoin/bitcoin/pull/18639#discussion_r408447328):
> Removed locks are generally replaced with `AssertLockHeld(cs_main);` and the corresponding compiler annotations

hebasto committed:
Prevent consecutive cs_main locks in UnloadBlockIndex

hebasto committed:
Prevent consecutive cs_main locks in CChainState::ReplayBlocks

hebasto commented:
Updated fe187562ef9765942d7c7d801e04d70d294d5aef  -> 18f0cb80c0cea7c9bc0b69a0f1fe0ca6cae8cde1 ([pr18639.02](https://github.com/hebasto/bitcoin/commits/pr18639.02) -> [pr18639.03](https://github.com/hebasto/bitcoin/commits/pr18639.03), [diff](https://github.com/hebasto/bitcoin/compare/pr18639.02..pr18639.03)):

analogous changes added for:
- `UnloadBlockIndex()`
- `CChainState::ReplayBlocks()`

MarcoFalke commented:
What problem is this solving?

hebasto commented:
@MarcoFalke 
> What problem is this solving?

Currently, `::cs_main` has type `RecursiveMutex` which derives from `std::recursive_mutex`.

But I'm agree with the following `TODO` statement:
https://github.com/bitcoin/bitcoin/blob/4a71c469058b824ed6c5132ab9901e194fa8aae1/src/sync.h#L106-L110

MarcoFalke commented:
Ok, Concept ACK based on that motivation, but OP should probably mention that. Some general notes, though:

* I am not sure if we can ever make `cs_main` not recursive. There might be circular dependencies in similar ways to how `csPathCached` can not be made a Mutex right now.
* We currently have no mechanism to prevent double locking a Mutex. This is undefined behavior and currently happens on some of the Mutexes we have. Someone should probably fix that.

hebasto commented:
>* We currently have no mechanism to prevent double locking a Mutex. This is undefined behavior and currently happens on some of the Mutexes we have. Someone should probably fix that.

Thanks for reminding :)

MarcoFalke commented:
tsan should detect those: https://github.com/llvm/llvm-project/blame/master/compiler-rt/test/tsan/mutex_double_lock.cpp

promag commented:
Code review ACK e09b98605f6fdd2ad3bcdf4c68d229905f7a29cd.

I don't see much value in this change especially because a) it's touching validation b) it's not fixing a problem.
